pyvo/utils/prototype.py

```diff
@@ -0,0 +1,191 @@
+import inspect
+import warnings
+from dataclasses import dataclass
+from functools import wraps
+from typing import Dict, Iterable
+
+from pyvo.dal.exceptions import PyvoUserWarning
+
+__all__ = ['features', 'prototype_feature', 'activate_features', 'Feature', 'PrototypeWarning', 'PrototypeError']
+
+features: Dict[str, "Feature"] = {
+
+}
+
+
+def prototype_feature(*args):
+    """
+    Decorator for functions and classes that implement unstable standards which haven't been approved yet.
+    The decorator can be used to tag individual functions or methods::
+
+       @prototype_feature('a-feature')
+       def i_am_a_prototype(*arg, **kwargs):
+           pass
+
+    In this case, a single function or method is tagged as part of the ``a-feature`` prototype feature. If the feature
+    has a URL defined (see :ref:`pyvo-prototypes-registry`).
+
+    Alternatively, a class can be marked as belonging to a feature. All public methods will be marked as part of the
+    prototype implementation. Protected, private, and *dunder* methods (i.e. any method starting with
+    an underscore) will be ignored. The reason is that the class might be instantiated by some mediator before the
+    user can call (and more importantly not call) a higher level facade::
+
+      @prototype_feature('a-feature')
+      class Feature:
```

Comment:
This is tough not the `Feature` class from below, right? If yes, please rename Feature here to something prototype-y.